flink-runtime/src/main/java/org/apache/flink/runtime/state/DefaultOperatorStateBackend.java

```diff
@@ -131,59 +169,134 @@ public void dispose() {
 
-			OperatorBackendSerializationProxy backendSerializationProxy =
-					new OperatorBackendSerializationProxy(metaInfoList);
+						for (Map.Entry<String, PartitionableListState<?>> entry :
+								registeredStatesDeepCopies.entrySet()) {
 
-			backendSerializationProxy.write(dov);
+							PartitionableListState<?> state = entry.getValue();
+							OperatorBackendSerializationProxy.StateMetaInfo<?> metaInfo =
+									new OperatorBackendSerializationProxy.StateMetaInfo<>(
+											state.getName(),
+											state.getPartitionStateSerializer(),
+											state.getAssignmentMode());
+							metaInfoList.add(metaInfo);
+						}
 
-			dov.writeInt(registeredStates.size());
-			for (Map.Entry<String, PartitionableListState<?>> entry : registeredStates.entrySet()) {
+						CheckpointStreamFactory.CheckpointStateOutputStream out = getIoHandle();
+						DataOutputView dov = new DataOutputViewStreamWrapper(out);
 
-				PartitionableListState<?> value = entry.getValue();
-				long[] partitionOffsets = value.write(out);
-				OperatorStateHandle.Mode mode = value.getAssignmentMode();
-				writtenStatesMetaData.put(entry.getKey(), new OperatorStateHandle.StateMetaInfo(partitionOffsets, mode));
-			}
+						OperatorBackendSerializationProxy backendSerializationProxy =
+								new OperatorBackendSerializationProxy(metaInfoList);
 
-			OperatorStateHandle handle = new OperatorStateHandle(writtenStatesMetaData, out.closeAndGetHandle());
+						backendSerializationProxy.write(dov);
 
-			return new DoneFuture<>(handle);
-		} finally {
-			closeStreamOnCancelRegistry.unregisterClosable(out);
-			out.close();
+						dov.writeInt(registeredStatesDeepCopies.size());
+
+						for (Map.Entry<String, PartitionableListState<?>> entry :
+								registeredStatesDeepCopies.entrySet()) {
+
+							PartitionableListState<?> value = entry.getValue();
+							long[] partitionOffsets = value.write(out);
+							OperatorStateHandle.Mode mode = value.getAssignmentMode();
+							writtenStatesMetaData.put(
+									entry.getKey(),
+									new OperatorStateHandle.StateMetaInfo(partitionOffsets, mode));
+						}
+
+						if (open.compareAndSet(true, false)) {
+
+							OperatorStateHandle operatorStateHandle =
+									new OperatorStateHandle(writtenStatesMetaData, out.closeAndGetHandle());
+
+							if (asynchronousSnapshots) {
+								LOG.info("DefaultOperatorStateBackend snapshot ({}, asynchronous part) in thread {} took {} ms.",
+										streamFactory, Thread.currentThread(), (System.currentTimeMillis() - asyncStartTime));
+							}
+
+							return operatorStateHandle;
+						} else {
+							throw new IOException("Checkpoint stream already closed.");
+						}
+					}
+
+					@Override
+					public void done(boolean canceled) {
+						if (open.compareAndSet(true, false)) {
+							CheckpointStreamFactory.CheckpointStateOutputStream stream = getIoHandle();
+							if (null != stream) {
+								closeStreamOnCancelRegistry.unregisterClosable(stream);
```

Comment:
Shouldn't we always try to remove the stream from the `ClosableRegistry` after the callable has been executed?